spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/reactor/ReactorSleuth.java

```diff
@@ -59,65 +57,69 @@ private ReactorSleuth() {
 	 * @param <T> an arbitrary type that is left unchanged by the span operator
 	 * @return a new lazy span operator pointcut
 	 */
-	@SuppressWarnings("unchecked")
 	public static <T> Function<? super Publisher<T>, ? extends Publisher<T>> scopePassingSpanOperator(
 			BeanFactory beanFactory) {
 		if (log.isTraceEnabled()) {
 			log.trace("Scope passing operator [" + beanFactory + "]");
 		}
 
+		// keep a reference outside the lambda so that any caching will be visible to
+		// all publishers
+		LazyCurrentTraceContext lazyCurrentTraceContext = new LazyCurrentTraceContext(
+				beanFactory);
+
 		return Operators.liftPublisher((p, sub) -> {
```

Comment:
note that the main difference between this and normal spring stuff, is usually there's an object to catch the reference of the bean we want to cache. since this is a lambda, there's no stash. I've extracted the stashing logic into `LazyCurrentTraceContext` although most of it is defensive ceremony around bean factories. the main advantage is we can pass the same type whether or not it is lazy.